### ISSUE TITLE
[FireMonkey] Protected header support

### ISSUE THREAD

**Issue by qsniyg** [erosman/support#100]
From looking at the docs, I'm not sure if this is out of scope for the project, but from looking at the source code, it appears the reason this is not supported is technical.

If it's just the latter, it's possible to support forbidden headers by using the webRequest API. I've implemented it in my extension here: https://github.com/qsniyg/maxurl/blob/master/extension/background.js (this is also the way Violentmonkey does it: [requests.js](https://github.com/violentmonkey/violentmonkey/blob/9e672d5590aea144840681b6f2ce0c267d57fc13/src/background/utils/requests.js) and [index.js](https://github.com/violentmonkey/violentmonkey/blob/9e672d5590aea144840681b6f2ce0c267d57fc13/src/common/index.js))

When doing a request, it renames all the headers to have a `IMU--` prefix (in your case, probably `FM--`), which allows the headers to go through. It also adds an `IMU-Verify` header (containing a random number) to ensure the request was indeed handled by the extension, instead of by the page (to avoid security issues), which is stored in a table.

Then it sets up an `onBeforeSendHeaders` listener that will then modify the headers, removing the `IMU--` prefix for each of the headers (and ensuring the value for the `IMU-Verify` header is indeed in the table). Afterwards, it removes the verification value from the table to ensure no memory is leaked.

This way, protected/forbidden headers can be used.

---

As for the use case, I've seen it used in many userscripts, and my userscript uses it extensively (close to a hundred times actually). Without protected headers, many sites (such as Instagram) will flat-out refuse to work.

**Comment by erosman** [comment]
I have made security a priority in FireMonkey. 

Under the "**Forbidden header name**" topic in Help, it lists the values of headers that can not be changed. The reason is obvious as they create security holes that can be exploited by malware scripts.
They are actually blocked by Firefox, not FireMonkey. In FireMonkey, I remove them in order to avoid errors popping up that would break the script when trying to change a forbidden header.

Are these the headers that you want to change?

**Comment by qsniyg** [comment]
>They are actually blocked by Firefox

Yes, that's why I posted the method above.

Using the method I explained in the issue, you can use forbidden headers under Firefox (or Chrome, etc.).

Try using Violentmonkey with this code snippet under Firefox:

```
// ==UserScript==
// @name New Script
// @namespace Violentmonkey Scripts
// @match *://*/*
// @grant GM_xmlhttpRequest
// @connect *
// ==/UserScript==

(function() {
  GM_xmlhttpRequest({
    method: "GET",
    url: "https://headers.cloxy.net/request.php",
    headers: {
      Referer: "http://google.com",
      "User-Agent": "fake useragent",
      "Cache-Control": "no-cache",
      "Pragma": "no-cache"
    },
    onload: function(resp) {
      console.log(resp.responseText);
    }
  })
})();
```

It'll work without issue (at least it does for me).

**Comment by erosman** [comment]
That can enable user-scripts to fake identity when connecting to users bank-account or other sensitive locations.
AMO may not allow such behaviour for user-script extensions which would result in extension being rejected.

I will seek advice from AMO first.

**Comment by qsniyg** [comment]
>That can enable user-scripts to fake identity when connecting to users bank-account or other sensitive locations.

Tampermonkey blacklists some banking sites by default (although I believe it's done as an example).

>AMO may not allow such behaviour for user-script extensions which would result in extension being rejected.

Possibly, but I'd personally be surprised. Greasemonkey, Tampermonkey, and Violentmonkey all implement this without issue from AMO.

**Comment by erosman** [comment]
I got the feedback from AMO security team allowing some header modifications. I will then implement the feature.

**Comment by erosman** [comment]
It is done and testing .... it will be in the next upgrade v1.18

**Comment by erosman** [comment]
v1.18 uploaded
